### ISSUE TITLE
"needs compatible type" as result of `ls | where size > 2kb` when running in directory "C:\Windows\System32"

### ISSUE THREAD

**Issue by adymorz** [nushell/nushell#4975]
### Describe the bug

Running  `ls | where size > 2kb` in `C:\Windows\System32` created the following error message:

```
C:\Windows\system32〉ls | where size > 2kb                                                        03/26/2022 09:49:11 AM
Error: nu::shell::type_mismatch (link)

  × Type mismatch
   ╭─[entry #18:1:1]
 1 │ ls | where size > 2kb
   ·                 ┬
   ·                 ╰── needs compatible type
   ╰────
```

### How to reproduce

1. Install nu shell over winget `winget install nushell`
2. Install windows terminal over winget `winget install Microsoft.WindowsTerminal`
3. Start windows terminal and add a new profile pointing to nu.exe
4. Start a new session using the nu profile
5. Go to "C:\Windows\System32"
6. Type `ls | where size > 2kb`

### Expected behavior

Expected to see a table with files larger than 2 kb

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.60.0                                   |
| short_commit       | 74af31a4                                 |
| commit_hash        | 74af31a42ff4feb8d5d9897c0d99c7434678a315 |
| commit_date        | 2022-03-22 19:07:11 +00:00               |
| build_os           | windows-x86_64                           |
| rust_version       | rustc 1.59.0 (9d1b2106e 2022-02-23)      |
| rust_channel       | stable-x86_64-pc-windows-msvc            |
| cargo_version      | cargo 1.59.0 (49d8809dc 2022-02-10)      |
| pkg_version        | 0.60.0                                   |
| build_time         | 2022-03-22 19:26:49 +00:00               |
| build_rust_channel | release                                  |
| features           | dataframe, default, trash, which, zip    |
| installed_plugins  |                                          |

**Comment by hustcer** [comment]
This issue could happen on macOS when I have other file type, say `socket`:
```
@hustcer ➜ tmp $> ls | first 3
  #             name               type       size       modified
─────────────────────────────────────────────────────────────────────
  0   !CloudShell                 dir           1,664   9 hours ago
  1   xxxxxxxxxPipe               socket   a week ago   ❎
  2   TemporaryDirectory.0FIBke   dir             128   9 hours ago

@hustcer ➜ tmp $> ls | first 3 | where size > 2kb
Error: nu::shell::type_mismatch (link)

  × Type mismatch
   ╭─[entry #31:1:1]
 1 │ ls | first 3 | where size > 2kb
   ·                           ┬
   ·                           ╰── needs compatible type
   ╰────
```

**Comment by adymorz** [comment]
The problem ist a folder called "DriverState" located under "C:\Windows\System32"

It is easy reproducible with `ls | where name == "DriverState" | where size > 2kb`

To find it out I used successive approximation running the command `ls | first 555 | last 5 | where size > 2kb` until I got a small enough amount to get the problematic entry.

**Comment by fdncred** [comment]
I bet that makes it a permissions issue. I'm not sure what the behavior should be when we don't have permission. Maybe just list the file name/folder name with no info?

**Comment by jntrnr** [comment]
I think that's how it currently works (well, worked in 0.44 and hopefully the same in 0.60). I'm guessing a bit, but I wonder if `size` field is coming back as missing, so it's either trying to compare an error with `2kb` or `$nothing` with `2kb`, and I don't think either are allowed.

**Comment by adymorz** [comment]
From a user perspective:

I am interested in a list of files whose size is above or below a threshold. It is not clear, what should be done with unkown sizes (which is the case on missing permissions).

So, it is not wrong what the command is doing.

As a user I could use the following command:

`ls | where size != $nothing && size > 2kb`

One possible improvement would be to show an explicit error message in case if a user uses the `>` operator on `$nothing`: `> operator is not defined for $nothing` (which tbh would improve the situation only a bit).

**Comment by hustcer** [comment]
BTW: It's weird that the table was rendered as (In /private/tmp dir on mac):
```
$> ls|last 5
  #                  name                    type       size       modified
──────────────────────────────────────────────────────────────────────────────
  0   mysqlx.sock                           socket   a week ago   ❎
  1   powerlog                              dir              64   2 days ago
  2   rustcwU3RnL                           dir              64   a day ago
  3   rustcyweB0q                           dir              64   a day ago
  4   {mircosoft-8798dcbb-9087-ab19-2110}   socket   a week ago   ❎
```
maybe should be rendered as:
```
  #                  name                    type       size       modified
──────────────────────────────────────────────────────────────────────────────
  0   mysqlx.sock                           socket        ❎      a week ago   
  1   powerlog                              dir              64   2 days ago
  2   rustcwU3RnL                           dir              64   a day ago
  3   rustcyweB0q                           dir              64   a day ago
  4   {mircosoft-8798dcbb-9087-ab19-2110}   socket        ❎      a week ago  
```

**Comment by fdncred** [comment]
I can see the same oddity in `/private/var/run`

**Comment by flying-sheep** [comment]
Somehow `into filesize` doesn’t seem to support null. or $nothing.

I came up with this workaround, but this should be fixed!

```nushell
[['a']; ['1'] ['']] | upsert a {|df| if ($df.a | empty?) { null } else { $df.a | into filesize } } 
```

**Comment by hustcer** [comment]
PR https://github.com/nushell/nushell/pull/5494 may help this issue especially on linux / mac, but not for permission related cases. What shall we do if the user doesn't have permissions?

**Comment by fdncred** [comment]
@hustcer I think it depends on the OS. This is what I get on Windows. For `size` i think we should return `0`. For the other stuff, I guess leave it `$nothing` until we can figure out how to get it to work like git-for-windows's `ls.exe` and windows' `cmd.exe`.
```
> ls -al c:\ | where type == $nothing 
╭──────────────────────────────┬──────┬────────┬──────┬─────────┬──────────┬──────────╮
│             name             │ type │ target │ size │ created │ accessed │ modified │
├──────────────────────────────┼──────┼────────┼──────┼─────────┼──────────┼──────────┤
│ c:\Config.Msi                │      │        │      │         │          │ ❎       │
│ c:\DumpStack.log             │      │        │      │         │          │ ❎       │
│ c:\DumpStack.log.tmp         │      │        │      │         │          │ ❎       │
│ c:\Recovery                  │      │        │      │         │          │ ❎       │
│ c:\System Volume Information │      │        │      │         │          │ ❎       │
│ c:\hiberfil.sys              │      │        │      │         │          │ ❎       │
│ c:\pagefile.sys              │      │        │      │         │          │ ❎       │
│ c:\swapfile.sys              │      │        │      │         │          │ ❎       │
├──────────────────────────────┼──────┼────────┼──────┼─────────┼──────────┼──────────┤
│             name             │ type │ target │ size │ created │ accessed │ modified │
╰──────────────────────────────┴──────┴────────┴──────┴─────────┴──────────┴──────────╯
```

`cmd.exe` clearly knows how to get information about these files.
```
04/08/2022  01:19 PM            12,288 DumpStack.log
05/06/2022  08:14 AM            12,288 DumpStack.log.tmp
05/06/2022  08:14 AM    13,618,864,128 hiberfil.sys
05/06/2022  08:14 AM     5,100,273,664 pagefile.sys
05/06/2022  08:14 AM        16,777,216 swapfile.sys
05/09/2022  09:18 AM    <DIR>          System Volume Information
10/27/2021  02:03 PM    <DIR>          Recovery
05/06/2022  01:13 PM    <DIR>          Config.Msi
```

`lsd.exe` produces this output.
```
lsd: c:\Config.Msi: Access is denied. (os error 5)
lsd: c:\DumpStack.log: Access is denied. (os error 5)
lsd: c:\DumpStack.log.tmp: The process cannot access the file because it is being used by another process. (os error 32)
lsd: c:\hiberfil.sys: The process cannot access the file because it is being used by another process. (os error 32)
lsd: c:\pagefile.sys: The process cannot access the file because it is being used by another process. (os error 32)
lsd: c:\Recovery: Access is denied. (os error 5)
lsd: c:\swapfile.sys: The process cannot access the file because it is being used by another process. (os error 32)
lsd: c:\System Volume Information: Access is denied. (os error 5)
```

**Comment by rgwood** [comment]
I am looking into this on Windows.

**Comment by rgwood** [comment]
OK, here's the situation as I currently understand it:

- We call [`std::fs::metadata()`](https://doc.rust-lang.org/stable/std/fs/fn.metadata.html) which calls `GetFileAttributesEx()` on Windows
- `GetFileAttributesEx()` is pretty fragile, it fails in many common scenarios
- The way the .NET runtime solves this: ["try again with FindFirstFile for any failure of GetFileAttributes"](https://github.com/dotnet/runtime/issues/26216#issuecomment-390269346)

We can probably fall back on [`FindFirstFile`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findfirstfilew) too. This may involve taking a dependency on the `windows` or `winapi` crates, which should be OK since we already build both as transitive dependencies.

**Comment by fdncred** [comment]
Good research. It would be good to do this or something like it. I hate returning empty columns when dir sees file sizes and files just fine, etc.

**Comment by rgwood** [comment]
I’ll aim to get this fixed on Windows for v0.64.

**Comment by rgwood** [comment]
Just merged a fix for this on Windows: #5703 

If all goes well, it will be released in v0.64 on June 14th.
